src/error/exporterrors.cpp

```diff
@@ -38,57 +38,32 @@ static ostream & printKDBErrors (ostream & os, parse_t & p)
-	   << "	using namespace ckdb;" << endl
-	   << endl
-	   << "#endif" << endl
-	   << "#define ELEKTRA_SET_ERROR(number, key, text) ELEKTRA_SET_ERROR_HELPER\\" << endl
-	   << "	(number, key, text, __FILE__, __LINE__)" << endl
-	   << endl
-	   << "#define ELEKTRA_SET_ERROR_HELPER(number, key, text, file, line) ELEKTRA_SET_ERROR_HELPER_HELPER\\" << endl
-	   << "	(number, key, text, file, line)" << endl
-	   << endl
-	   << "#define ELEKTRA_SET_ERROR_HELPER_HELPER(number, key, text, file, line) do {ELEKTRA_LOG (\"Add Error %d: %s\", "
-	      "number, text); elektraSetError ## number\\"
-	   << endl
-	   << "	(key, text, file, #line); } while (0)" << endl
-	   << endl
-	   << endl
-	   << "#define ELEKTRA_ADD_WARNING(number, key, text) ELEKTRA_ADD_WARNING_HELPER\\" << endl
-	   << "	(number, key, text, __FILE__, __LINE__)" << endl
-	   << "" << endl
-	   << "#define ELEKTRA_ADD_WARNING_HELPER(number, key, text, file, line) ELEKTRA_ADD_WARNING_HELPER_HELPER\\" << endl
-	   << "	(number, key, text, file, line)" << endl
-	   << "" << endl
-	   << "#define ELEKTRA_ADD_WARNING_HELPER_HELPER(number, key, text, file, line) do {ELEKTRA_LOG (\"Add Warning %d: %s\", "
-	      "number, text);  elektraAddWarning ## number\\"
-	   << endl
-	   << "	(key, text, file, #line);} while (0)" << endl
-	   << endl
-	   << endl
-	   << "#define ELEKTRA_SET_ERRORF(number, key, text, ...) ELEKTRA_SET_ERRORF_HELPER\\" << endl
-	   << "	(number, key, text, __FILE__, __LINE__, __VA_ARGS__)" << endl
-	   << endl
-	   << "#define ELEKTRA_SET_ERRORF_HELPER(number, key, text, file, line, ...) ELEKTRA_SET_ERRORF_HELPER_HELPER\\" << endl
-	   << "	(number, key, text, file, line, __VA_ARGS__)" << endl
-	   << endl
-	   << "#define ELEKTRA_SET_ERRORF_HELPER_HELPER(number, key, text, file, line, ...) do {ELEKTRA_LOG (\"Add Error \" "
-	      "ELEKTRA_STRINGIFY(number) \" : \" text, __VA_ARGS__); elektraSetErrorf ## number\\"
-	   << endl
-	   << "	(key, text, file, #line,  __VA_ARGS__); } while (0)" << endl
-	   << endl
-	   << endl
-	   << "#define ELEKTRA_ADD_WARNINGF(number, key, text, ...) ELEKTRA_ADD_WARNINGF_HELPER\\" << endl
-	   << "	(number, key, text, __FILE__, __LINE__, __VA_ARGS__)" << endl
-	   << "" << endl
-	   << "#define ELEKTRA_ADD_WARNINGF_HELPER(number, key, text, file, line, ...) ELEKTRA_ADD_WARNINGF_HELPER_HELPER\\" << endl
-	   << "	(number, key, text, file, line, __VA_ARGS__)" << endl
-	   << "" << endl
-	   << "#define ELEKTRA_ADD_WARNINGF_HELPER_HELPER(number, key, text, file, line, ...)  do {ELEKTRA_LOG (\"Add Warning \" "
-	      "ELEKTRA_STRINGIFY(number) \" : \" text, __VA_ARGS__); elektraAddWarningf ## number\\"
-	   << endl
-	   << "	(key, text, file, #line, __VA_ARGS__); } while (0)" << endl
-	   << endl
-	   << endl;
+	   << "using namespace ckdb;" << endl
+	   << "#endif" << endl;
+	os << endl << endl;
+
+	for (size_t i = 1; i < p.size (); ++i)
+	{
+		os << "#define ELEKTRA_SET_" << p[i]["macro"] << "_ERROR(key, text) \\" << endl
```

Comment:
Maybe it is too late now but actually it would make a lot of sense to rewrite this to use mustache (it is C++ anyway and would be much cleaner then).
@kodebach what do you think? Would it be easy to share the mustache header?

Reply:
> @kodebach what do you think?

I expected the new version of `exporterrors.cpp` to be much smaller, otherwise I would have suggested it already.

Since there the template is very static (only a few things change for each error), it should be fairly straightforward to switch to mustache. Although this could easily be done in a follow up PR.

> Would it be easy to share the mustache header?

We could simply copy it. Putting it into a common location would work, might be awkward since the only sensible location would be `src/include`.

Reply:
> Since there the template is very static (only a few things change for each error), it should be fairly straightforward to switch to mustache. Although this could easily be done in a follow up PR.

It is much easier to review mustache code than this print code.

>  We could simply copy it. Putting it into a common location would work, might be awkward since the only sensible location would be src/include.

Copying the copy is awkward. Maybe we create a src/include/external which contains header files we stole from other projects?

Reply:
> Maybe we create a src/include/external which contains header files we stole from other projects?

Maybe we should use `external/mustache/include` instead, i.e. one folder under `external` for each external dependency. Then it would also be easier to find and update external code. Right now some plugins (e.g. `ni`) also include source copies of libraries.

Reply:
In ni, the header file is not reused outside of the plugin.